File: golang/tests/pkg/integration/get_cdp_link_integration_test.go
package integration

import (
	"os"
	"strings"
	"testing"

	"github.com/alibabacloud-go/tea/dara"
	"github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/stretchr/testify/assert"
)

func TestGetCdpLink_WithBrowserSession(t *testing.T) {
	// Skip if no API key
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY environment variable not set")
	}

	// Create AgentBay client
	agentBayClient, err := agentbay.NewAgentBay(apiKey, nil)
	assert.NoError(t, err)

	// Create browser session
	params := agentbay.NewCreateSessionParams().WithImageId("browser_latest")
	result, err := agentBayClient.Create(params)
	assert.NoError(t, err)
	assert.NotNil(t, result)
	assert.True(t, result.Success)
	assert.NotNil(t, result.Session)

	session := result.Session
	defer session.Delete()

	// Test GetCdpLink
	request := &client.GetCdpLinkRequest{
		Authorization: dara.String("Bearer " + apiKey),
		SessionId:     dara.String(session.GetSessionID()),
	}

	response, err := agentBayClient.Client.GetCdpLink(request)
	
	// Check if API is available
	if err != nil && strings.Contains(err.Error(), "InvalidAction.NotFound") {
		t.Skip("GetCdpLink API not yet available in production")
		return
	}
	
	assert.NoError(t, err)
	assert.NotNil(t, response)
	assert.NotNil(t, response.Body)
	assert.True(t, *response.Body.Success)
	assert.NotNil(t, response.Body.Data)
	assert.NotNil(t, response.Body.Data.Url)

	url := *response.Body.Data.Url
	assert.True(t, strings.HasPrefix(url, "ws://") || strings.HasPrefix(url, "wss://"),
		"CDP URL should start with ws:// or wss://")
	
	t.Logf("CDP URL: %s", url)
}

func TestGetCdpLink_WithInvalidSession(t *testing.T) {
	// Skip if no API key
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY environment variable not set")
	}

	// Create AgentBay client
	agentBayClient, err := agentbay.NewAgentBay(apiKey, nil)
	assert.NoError(t, err)

	// Test GetCdpLink with invalid session
	request := &client.GetCdpLinkRequest{
		Authorization: dara.String("Bearer " + apiKey),
		SessionId:     dara.String("invalid-session-id-12345"),
	}

	_, err = agentBayClient.Client.GetCdpLink(request)
	
	// Should get an error for invalid session
	assert.Error(t, err)
	assert.True(t, 
		strings.Contains(err.Error(), "InvalidMcpSession.NotFound") ||
		strings.Contains(err.Error(), "InvalidAction.NotFound"),
		"Expected InvalidMcpSession.NotFound or InvalidAction.NotFound error")
}

File: typescript/tests/integration/get-cdp-link.integration.test.ts
import { AgentBay } from "../../src/agent-bay";
import { CreateSessionParams } from "../../src/session-params";
import { GetCdpLinkRequest } from "../../src/api/models/model";

describe("GetCdpLink Integration Test", () => {
  let agentBay: AgentBay;

  beforeAll(() => {
    const apiKey = process.env.AGENTBAY_API_KEY;
    if (!apiKey) {
      throw new Error("AGENTBAY_API_KEY environment variable not set");
    }
    agentBay = new AgentBay({ apiKey });
  });

  it("should get CDP link with a real browser session", async () => {
    const params = new CreateSessionParams();
    params.imageId = "linux_latest";
    const sessionResult = await agentBay.create(params);

    expect(sessionResult).not.toBeNull();
    expect(sessionResult.success).toBe(true);
    expect(sessionResult.session).not.toBeNull();

    const session = sessionResult.session!;

    try {
      const request = new GetCdpLinkRequest({
        authorization: `Bearer ${agentBay.apiKey}`,
        sessionId: session.sessionId,
      });

      let response;
      try {
        response = await agentBay.client.getCdpLink(request);
      } catch (error: any) {
        if (error.message && error.message.includes("InvalidAction.NotFound")) {
          console.log("Skipping test: GetCdpLink API not yet available in production");
          return;
        }
        throw error;
      }

      expect(response).not.toBeNull();
      expect(response.body).not.toBeNull();
      expect(response.body?.success).toBe(true);
      expect(response.body?.data).not.toBeNull();
      expect(response.body?.data?.url).not.toBeNull();

      const url = response.body?.data?.url;
      expect(url?.startsWith("ws://") || url?.startsWith("wss://")).toBe(true);
      console.log(`CDP URL: ${url}`);
    } finally {
      await session.delete();
    }
  }, 60000);

  it("should throw error with invalid session ID", async () => {
    const request = new GetCdpLinkRequest({
      authorization: `Bearer ${agentBay.apiKey}`,
      sessionId: "invalid-session-id-12345",
    });

    try {
      await agentBay.client.getCdpLink(request);
      fail("Should have thrown an error");
    } catch (error: any) {
      expect(
        error.message.includes("InvalidMcpSession.NotFound") ||
          error.message.includes("InvalidAction.NotFound")
      ).toBe(true);
    }
  }, 30000);
});

